src/main/java/appeng/parts/reporting/PartConversionMonitor.java

```diff
@@ -89,62 +91,133 @@ public boolean onPartShiftActivate( final EntityPlayer player, final EnumHand ha
 			return false;
 		}
 
-		boolean ModeB = false;
+		final ItemStack eq = player.getHeldItem( hand );
+		if( this.isLocked() )
+		{
+			if( eq.isEmpty() )
+			{
+				this.insertItem( player, hand, true );
+			}
+			else if( Platform.isWrench( player, eq, this.getLocation().getPos() ) && ( this.getDisplayed() == null || !this.getDisplayed().equals( eq ) ) )
+			{
+				// wrench it
+				return super.onPartActivate( player, hand, pos );
+			}
+			else
+			{
+				this.insertItem( player, hand, false );
+			}
+		}
+		else if( this.getDisplayed() != null && this.getDisplayed().equals( eq ) )
+		{
+			this.insertItem( player, hand, false );
+		}
+		else
+		{
+			return super.onPartActivate( player, hand, pos );
+		}
+
+		return true;
+	}
 
-		ItemStack item = player.getHeldItem( hand );
-		if( item.isEmpty() && this.getDisplayed() != null )
+	@Override
+	public boolean onClicked( EntityPlayer player, EnumHand hand, Vec3d pos )
+	{
+		if( Platform.isClient() )
 		{
-			ModeB = true;
-			item = ( (IAEItemStack) this.getDisplayed() ).createItemStack();
+			return true;
 		}
 
-		if( !item.isEmpty() )
+		if( !this.getProxy().isActive() )
 		{
-			try
-			{
-				if( !this.getProxy().isActive() )
-				{
-					return false;
-				}
+			return false;
+		}
 
-				final IEnergySource energy = this.getProxy().getEnergy();
-				final IMEMonitor<IAEItemStack> cell = this.getProxy().getStorage().getInventory(
-						AEApi.instance().storage().getStorageChannel( IItemStorageChannel.class ) );
-				final IAEItemStack input = AEItemStack.fromItemStack( item );
+		if( !Platform.hasPermissions( this.getLocation(), player ) )
+		{
+			return false;
+		}
 
-				if( ModeB )
+		if( this.getDisplayed() != null )
+		{
+			this.extractItem( player, this.getDisplayed().getDefinition().getMaxStackSize() );
+		}
+
+		return true;
+	}
+
+	@Override
+	public boolean onShiftClicked( EntityPlayer player, EnumHand hand, Vec3d pos )
+	{
+		if( Platform.isClient() )
+		{
+			return true;
+		}
+
+		if( !this.getProxy().isActive() )
+		{
+			return false;
+		}
+
+		if( !Platform.hasPermissions( this.getLocation(), player ) )
+		{
+			return false;
+		}
+
+		if( this.getDisplayed() != null )
+		{
+			this.extractItem( player, 1 );
+		}
+
+		return true;
+	}
+
+	private void insertItem( final EntityPlayer player, final EnumHand hand, final boolean allItems )
+	{
+		try
+		{
+			final IEnergySource energy = this.getProxy().getEnergy();
+			final IMEMonitor<IAEItemStack> cell = this.getProxy()
+					.getStorage()
+					.getInventory(
+							AEApi.instance().storage().getStorageChannel( IItemStorageChannel.class ) );
+			if( allItems )
+			{
+				final IAEItemStack input = this.getDisplayed().copy();
+				IItemHandler inv = new PlayerMainInvWrapper( player.inventory );
+
+				for( int x = 0; x < inv.getSlots(); x++ )
 				{
-					for( int x = 0; x < player.inventory.getSizeInventory(); x++ )
+					final ItemStack targetStack = inv.getStackInSlot( x );
+					if( input.equals( targetStack ) )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this call to "equals"; comparisons between unrelated types always return false. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2159)